source/common/websocket/codec.cc

```diff
@@ -0,0 +1,219 @@
+    return absl::nullopt;
+  }
+  std::vector<uint8_t> output;
+  // Set flags and opcode
+  pushScalarToByteVector(
+      static_cast<uint8_t>(frame.final_fragment_ ? (0x80 | frame.opcode_) : frame.opcode_), output);
+
+  // Set payload length
+  if (frame.payload_length_ <= 125) {
+    // Set mask bit and 7-bit length
+    pushScalarToByteVector(frame.masking_key_ ? static_cast<uint8_t>(frame.payload_length_ | 0x80)
+                                              : static_cast<uint8_t>(frame.payload_length_),
+                           output);
+  } else if (frame.payload_length_ <= 65535) {
+    // Set mask bit and 16-bit length indicator
+    pushScalarToByteVector(static_cast<uint8_t>(frame.masking_key_ ? 0xfe : 0x7e), output);
+    // Set 16-bit length
+    pushScalarToByteVector(htobe16(frame.payload_length_), output);
+  } else {
+    // Set mask bit and 64-bit length indicator
+    pushScalarToByteVector(static_cast<uint8_t>(frame.masking_key_ ? 0xff : 0x7f), output);
+    // Set 64-bit length
+    pushScalarToByteVector(htobe64(frame.payload_length_), output);
+  }
+  // Set masking key
+  if (frame.masking_key_) {
+    pushScalarToByteVector(htobe32(frame.masking_key_.value()), output);
+  }
+  return output;
+}
+
+void Decoder::frameMaskFlag(uint8_t mask_and_length) {
+  num_remaining_masking_key_bytes_ = mask_and_length & 0x80 ? kMaskingKeyLength : 0;
+  length_ = mask_and_length & 0x7F;
+}
+
+void Decoder::frameDataStart() {
+  frame_.payload_length_ = length_;
+  if (length_ == 0) {
+    state_ = State::FrameFinished;
+  } else {
+    if (decode_payload_) {
+      frame_.payload_ = std::make_unique<Buffer::OwnedImpl>();
+    }
+    state_ = State::FramePayload;
+  }
+}
+
+void Decoder::frameData(const uint8_t* mem, uint64_t length) { frame_.payload_->add(mem, length); }
+
+void Decoder::frameDataEnd(absl::optional<std::vector<Frame>>& output) {
+  output->push_back(std::move(frame_));
+  resetDecoder();
+}
+
+void Decoder::resetDecoder() {
+  frame_ = {false, 0, absl::nullopt, 0, nullptr};
+  state_ = State::FrameHeaderFlagsAndOpcode;
+  length_ = 0;
+  num_remaining_extended_length_bytes_ = 0;
+  num_remaining_masking_key_bytes_ = 0;
+}
+
+uint8_t Decoder::doDecodeFlagsAndOpcode(absl::Span<const uint8_t>& data) {
+  // Validate opcode (last 4 bits)
+  uint8_t opcode = data.front() & 0x0f;
+  if (std::find(kFrameOpcodes.begin(), kFrameOpcodes.end(), opcode) == kFrameOpcodes.end()) {
+    ENVOY_LOG(debug, "Failed to decode websocket frame with invalid opcode: {}", opcode);
+    return 0;
+  }
+  frame_.opcode_ = opcode;
+  frame_.final_fragment_ = data.front() & 0x80;
+  state_ = State::FrameHeaderMaskFlagAndLength;
+  return 1;
+}
+
+uint8_t Decoder::doDecodeMaskFlagAndLength(absl::Span<const uint8_t>& data) {
+  frameMaskFlag(data.front());
+  if (length_ == 0x7e) {
+    num_remaining_extended_length_bytes_ = kPayloadLength16Bit;
+    length_ = 0;
```

Comment:
Tell the true, the `length_ = 0` make me confused at first glance. I think we can simply merge `frameMaskFlag` to this function and set a correct `state_` based on the flag/length indicator value.

Reply:
If `length_` 7 bit value <= 253 then it is the length. If it is 254 or 255 then it is a length indicator and we reset it to zero to read the actual value with the bytes we read in. I added a comment to make it more clear. `frameMaskFlag` is merged to the caller.